pkg/reconciler/labeler/v2/accessors.go

```diff
@@ -57,9 +59,48 @@ func NewRevisionAccessor(
 	}
 }
 
-// get implements Accessor
-func (r *Revision) get(ns, name string) (kmeta.Accessor, error) {
-	return r.revisionLister.Revisions(ns).Get(name)
+// makeMetadataPatch makes a metadata map to be patched or nil if no changes are needed.
+func makeMetadataPatch(acc kmeta.Accessor, routeName *string) (map[string]interface{}, error) {
+	labels := map[string]interface{}{}
+
+	if err := addRouteLabel(acc, labels, routeName); err != nil {
+		return nil, err
+	}
+
+	if len(labels) > 0 {
+		meta := map[string]interface{}{
+			"metadata": map[string]interface{}{
+				"labels": labels,
+			},
+		}
+		return meta, nil
+	}
+	return nil, nil
+}
+
+// addRouteLabel appends the route label to the list of labels if needed
+// or removes the label if routeName is nil.
+func addRouteLabel(acc kmeta.Accessor, diffLabels map[string]interface{}, routeName *string) error {
+	oldLabels := acc.GetLabels()
+	if routeName == nil { // remove the label
+		if len(oldLabels) != 0 && oldLabels[serving.RouteLabelKey] != "" {
+			diffLabels[serving.RouteLabelKey] = nil
+		}
+	} else { // add the label
+		if len(oldLabels) == 0 {
+			diffLabels[serving.RouteLabelKey] = routeName
```

Comment:
we want `routeName` or `*routeName` here?

Reply:
I think we want *routeName so its a string -> string comparison
It's a pointer so we can remove with a nil pointer

Reply:
then again I wonder if we can get rid of this whole pointer and treat empty string as 'remove'